### PR TITLE
Sphere occluders (portals and general use)

### PR DESCRIPTION
**Pull request by lawnjelly**
Add framework for supporting geometrical occluders, and add support for sphere occluders.
Includes gizmos for editing.
Occluders now work both within the portal system AND also in general use in scenes with no room system (no RoomManager / Rooms etc).

## Geometric Occluders
Now the framework for rooms & portals is built, it is quite easy to add support for geometric occluders. Essentially when culling objects in a room, we can go through a list of occluding primitives, and see if the objects are hidden behind these. It turns out these can also be used for simple occlusion even without the portal system.

These occluders aren't quite as powerful as portals themselves, but can help in situations where it is difficult to directly use portals. In particular they are handy for:

* Terrain
* Dynamic occluders (you can e.g. add one or several occluders to a spaceship sub-scene)
* Occlusion in levels that have not been portalled (including automatic)

occluder_spheres2

They will occlude objects and also portals, so if e.g. a cave entrance portal is hidden behind a mountain, it will not be rendered, and none of the cave.

I've started with the simplest, spheres. But after this is should be possible to add other occluder primitive shape types, including anti-portals (convex occluding polys), and possibly merged SDF primitives (like metaballs). Anti-portals could be useful because they can be quite easily automatically placed in levels, to gain some limited occlusion culling benefit even without manually splitting into rooms.

There are a few disadvantages of geometric occluders:
* They can't easily be 'merged' at runtime. If there are e.g. two spheres, an object will only be culled if it is completely behind one of the spheres. If it is half obscured by one, and half by the other, it will not be culled. This contrasts with the raster occlusion culling in 4.x which _does_ deal with this merging problem.
* This is purely for occlusion culling, it cannot be used for gameplay notifications.

I've gone to some lengths to ensure that these occluders are *almost* free to use.
* Multiple occluders can be contained within a single RoomOccluder node. This means the node overhead is very low, and you can easily create hundreds if you desire.
* At runtime, a preparation phase preselects the most important occluders based on screen space metric (size and distance from camera). So you may have e.g. 100 occluders, but it can choose e.g. the 8 most valid, and only test the objects against these.

## Notes
* These occluders don't have to be placed in rooms, or in the portal system. They will automatically attach to the system when running, much like roaming objects, and can be created and deleted at any time. This is necessary as users will often add them to e.g. spaceships which may be roaming objects.
* When in use without the portal system, they will not cull gizmos (it still checks and respects the GLOBAL portal_mode even when portals are off).
* These can probably be optimized a bit more especially for large rooms or general use, but should be pretty good until hundreds of occluders are in use.

### CODE REVIEW

**Comment by Calinou** [comment]
This is a pretty good idea, as it will allow for easier upgrades to Godot 4.0's raster-based occlusion culling.

**Comment by lawnjelly** [comment]
> This is a pretty good idea, as it will allow for easier upgrades to Godot 4.0's raster-based occlusion culling. It will still require some kind of upgrade tool, but it sounds feasible to do.

Yes we can have a look at the details of the raster occlusion, see if we can share some setup, certainly I'm thinking of a similar idea in terms of automated baking of large static occluders for the anti-portals. Even in 4.x we might end up having a geometric option available eventually as an alternative, depending on how performant the raster solution is on mobile.

> Is it possible to move those occluder spheres every frame without causing a performance loss? This would be useful for large dynamic objects that may obstruct the player's vision.

Yeah essentially dynamic spheres are just a vertex transform. So with multiple spheres in a node they can be transformed efficiently every frame. Aside from that the only upkeep it to keep track of which room the overall multi-occluder is in, and that's fairly cheap (like a roaming object). 

> Also, is non-uniform scaling supported? If scaling won't work as expected, there should probably be a node configuration warning.

As the ray sphere test is done in world space the scaling is applied uniformly on all axes. Yep I'll add a node configuration warning that's a good idea.

File: doc/classes/EditorSpatialGizmoPlugin.xml [hunk 1]
```diff
@@ -44,9 +44,9 @@
 		<method name="create_handle_material">
 			<return type="void" />
 			<argument index="0" name="name" type="String" />
-			<argument index="1" name="billboard" type="bool" default="false" />
+			<argument index="1" name="billboard" type="bool" />
+			<argument index="2" name="arg2" type="Texture" default="false" />
 			<description>
-				Creates a handle material with its variants (selected and/or editable) and adds them to the internal material list. They can then be accessed with [method get_material] and used in [method EditorSpatialGizmo.add_handles]. Should not be overridden.
```
Comment (akien-mga) [inline doc/classes/EditorSpatialGizmoPlugin.xml:49]: You lost the documentation here.

File: doc/classes/EditorSpatialGizmoPlugin.xml [hunk 2]
```diff
@@ -44,9 +44,9 @@
 		<method name="create_handle_material">
 			<return type="void" />
 			<argument index="0" name="name" type="String" />
-			<argument index="1" name="billboard" type="bool" default="false" />
+			<argument index="1" name="billboard" type="bool" />
+			<argument index="2" name="arg2" type="Texture" default="false" />
```
Comment (akien-mga) [inline doc/classes/EditorSpatialGizmoPlugin.xml:48]: Argument name missing and default value is wrong.

**Comment by lawnjelly** [comment]
Getting there on this. I would like to check whether there aren't any preferred renamings. Currently the names are as follows:

* `RoomOccluder` - this is the main node in the scene graph. It could also potentially just be called `Occluder` .. I know there is an Occluder2D already. In Godot 4.x `OccluderInstance3D` is used, but we could always rename this if ported to 4.x.
* `OccluderShape` - base class for the Shape resource. Note that a shape can potentially include multiple primitives, so maybe that could influence naming...
* `OccluderShape_Sphere` - I know some are not so keen on underscores so I'd welcome ways to name this, given that we will likely have at least a couple more `OccluderShape` types.

The gizmo lines are now magenta, they didn't show up very well in the earlier screenshot. They are customizable in the editor settings.

One thing that might be better is if the gizmo lines were drawn on top of other geometry, given that often you want to place occluders *inside* objects. But I couldn't manage to get this working despite setting as many `on top` flags as I could find... Is this possible or is it only for the billboard icons?

**Comment by Calinou** [comment]
> OccluderShape_Sphere - I know some are not so keen on underscores so I'd welcome ways to name this, given that we will likely have at least a couple more OccluderShape types.

I'd name it SphereOccluderShape.

**Comment by Calinou** [comment]
> One thing that might be better is if the gizmo lines were drawn on top of other geometry, given that often you want to place occluders inside objects. But I couldn't manage to get this working despite setting as many on top flags as I could find... Is this possible or is it only for the billboard icons?

This is set by the gizmo icon status in the 3D viewport's **View > Gizmos** menu.

If you click the icon twice, it will enter X-ray mode, which makes the gizmo's lines visible through walls.

**Comment by lawnjelly** [comment]
> This is set by the gizmo icon status in the 3D viewport's **View > Gizmos** menu.

Great! That works fab.

It just occurred to me that one of the reasons I was not sure about whether to use the node name `RoomOccluder`, is that although now they are tied to room system, there is a possibility in the future I could get them working even when a room system was not present, and have them available as a generalized simple occluder system. If that were the case then the name `RoomOccluder` would no longer make sense. What do you guys think?

**Comment by lawnjelly** [comment]
I've pretty much come round to the opinion that it would be unwise to prematurely tie ourselves in to the room system for the naming of the occluders, as it is very possible we can get them working even outside of rooms, i.e. just add them to general scenes and have them work.

With a lack of feedback so far I've done a search and replace type second commit to change the name from `RoomOccluder` to simply `Occluder`. But I would welcome any other ideas.

Some considerations:
* We could try and use a name for compatibility with the existing name `OcclusionInstance3D` in 4.x, with the idea of keeping the name if we add this to 4.x. However in practice users will have to upgrade the project anyway, so there doesn't seem to be a strong case to necessitate the same naming.
* I did also consider `GeomOccluder`, given that these are all geometrical occluders (as opposed to raster occluders). But given the lack of raster occlusion in 3.x, I'm not sure we would need to add the prefix.
* Whether or not to add a `3D`. In 3.x the convention seems mostly to omit the postfix for 3D nodes, and add the postfix for 2D nodes. So even though this has changed in 4.x, to have some nodes follow this convention and some not in 3.x might be a bit confusing.

I am in no way committed to any particular naming, and it is not that involved to change, so feel free to make any suggestions on this front. :+1: Once we have settled on naming I can squash any commits.

**Comment by Calinou** [comment]
> I did also consider GeomOccluder, given that these are all geometrical occluders (as opposed to raster occluders). But given the lack of raster occlusion in 3.x, I'm not sure we would need to add the prefix.

I think naming it Occluder is less confusing, as people might assume that occlusion culling won't work on particles and lights if the name GeomOccluder (or GeometryOccluder) is used.

**Comment by Calinou** [comment]
I ported the occluder icons from `master` to `3.x` with decreased saturation to match:

[icons.zip](https://github.com/godotengine/godot/files/6984091/icons.zip)

image

image

(We generally don't add icons to abstract classes.)

**Comment by lawnjelly** [comment]
Fantastic! Thanks, I will add them. :+1: :grinning:

**Comment by timothyqiu** [comment]
Got an error when moving the `Occluder` node or pressing the "Center Node" button:

```
 servers/visual/visual_server_scene.cpp:1203 - Condition "!ro->scenario" is true.
```

1. Add an `Occluder` node
2. Create an `OccluderShapeSphere` resource for it
3. Move the `Occluder` node around in the scene, or click the "Center Node" button on the toolbar

The error will be gone if I switch to another scene and then switch back, or reparent the node in the scene tree.

**Comment by lawnjelly** [comment]
@timothyqiu I'll check that error out it sounds like an order of construction thing. :+1: 

Ah that's good with the user interface feedback, good points.

* I'll try and add a default radius (as this is done by adding points in the Inspector I might have to just scan the new points and check for a 0.0 but it can work).
* I had no idea holding shift while dragging worked :grin: (must be a feature of secondary handles for exactly this kind of thing)
* Although having the radius bigger enables setting it from the UI from zero, I think there is an advantage to keeping the bigger one for moving, as that is the more common operation, and it could get fiddly with a smaller handle. Maybe we can set a minimum radius to help with this.

**Comment by timothyqiu** [comment]
The previous error is gone 😄

But now everytime I open a scene with Occluder + OccluderShapeSphere, there will be an error in the output:

```
scene/3d/spatial.cpp:504 - Condition "!is_inside_world()" is true. Returned: Ref<World>()
```

**Comment by lawnjelly** [comment]
> The previous error is gone 
> 
> But now everytime I open a scene with Occluder + OccluderShapeSphere, there will be an error in the output:
> 
> ```
> scene/3d/spatial.cpp:504 - Condition "!is_inside_world()" is true. Returned: Ref<World>()
> ```

Oh darn it lol! :grinning: Ok will fix!

I've set the default radius of new spheres to 1.0 which should make editing much easier.

Note that because there is no specific call to tell the resource the editor is creating a new data point, I have some basic logic which just tries to detect this, and there is a remote possibility it could convert an intentional 0.0 radius sphere to 1.0 at some point. But this is pretty remote and I spheres of size 0.0 are of no use anyway. I could possibly prevent intentional 0.0 size, or put a lower limit, but it's difficult to know in advance the scale the world users will use.

**Comment by lawnjelly** [comment]
Do let me know any more error messages I might have missed when using the spheres.

Aside from that I'm of the view it would be better to get this merged sooner than later so it gets a couple of days internal testing if @akien-mga is still intending to make a beta before going on holiday (I don't know if that will be possible due to octahedral / split stream mesh bugs).

It should be fairly risk free either way, as the new functionality is only active when occluders are present.

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**